components/enrichers/reachability/internal/atom/purl/purl.go

```diff
@@ -0,0 +1,143 @@
+package purl
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+// Parser allows to extract information from purls - https://github.com/package-url/purl-spec.
+type Parser struct {
+	matcherPurlPkg             *regexp.Regexp
+	matcherPurlTrailingVersion *regexp.Regexp
+	matcherPurlVersion         *regexp.Regexp
+}
+
+func NewParser() (*Parser, error) {
+	purlPkg, err := regexp.Compile(`(?P<p1>[^/:]+/(?P<p2>[^/]+))(?:(?:.|/)v\d+)?@`)
+	if err != nil {
+		return nil, fmt.Errorf("failed to compile purl pkg regex: %w", err)
+	}
+	purlTrailingVersion, err := regexp.Compile(`[./]v\d+@`)
+	if err != nil {
+		return nil, fmt.Errorf("failed to compile purl trailing version regex: %w", err)
+	}
+	purlVersion, err := regexp.Compile(`@(?P<v1>v?(?P<v2>[\d.]+){1,3})(?P<ext>[^?\s]+)?`)
+	if err != nil {
+		return nil, fmt.Errorf("failed to compile purl version regex: %w", err)
+	}
+
+	return &Parser{
+		matcherPurlPkg:             purlPkg,
+		matcherPurlTrailingVersion: purlTrailingVersion,
+		matcherPurlVersion:         purlVersion,
+	}, nil
+}
+
+// ParsePurl extracts pkg:version matches from the supplied purl.
+func (p *Parser) ParsePurl(purl string) ([]string, error) {
+	purl = p.matcherPurlTrailingVersion.ReplaceAllString(purl, "$1@")
```

Comment:
I'm refactoring this to use `packageurl` but need some extra time to make sure that the refactor actually makes sense

From my findings, the actual code spits out the following:

```go
func TestParser_ParsePurl(t *testing.T) {
	pParser, err := purl.NewParser()
	require.NoError(t, err)

	for _, tt := range []struct {
		inputPurl       string
		expectedMatches []string
	}{
		{
			inputPurl: "pkg:bitbucket/birkenfeld/pygments-main@244fd47e07d1014f0aed9c",
			expectedMatches: []string{
				"birkenfeld/pygments-main:244",
				"birkenfeld/pygments-main:244fd47e07d1014f0aed9c",
				"pygments-main:244",
				"pygments-main:244fd47e07d1014f0aed9c",
			},
		},
	} {
		t.Run(fmt.Sprintf("it should correctly parse with input %s", tt.inputPurl), func(t *testing.T) {
			pp, err := pParser.ParsePurl(tt.inputPurl)
			require.NoError(t, err)
			assert.Equal(t, pp, tt.expectedMatches)
		})
	}
}
```

And I was able to refactor is as:

```go
func (p *Parser) ParsePurl(purl string) ([]string, error) {
	pp, err := packageurl.FromString(purl)
	if err != nil {
		return nil, fmt.Errorf("failed to parse purl: %w", err)
	}

	var (
		namespace    = pp.Namespace
		name         = pp.Name
		version      = pp.Version
		shortVersion string
	)

	if len(version) >= 3 {
		shortVersion = version[:3]
	}

	return []string{
		path.Join(namespace, name) + ":" + shortVersion,
		path.Join(namespace, name),
	}, nil
}
```

---

To me the arbitrary choices of extracting partial versions are bizarre - will try to make more sense of them

Feel free to share your thoughts on it @northdpole

Reply:
they are bizare indeed, i think it's the specific case where the version is a commit sha however

Reply:
I'll raise a smaller PR to improve this bit, if that's ok!